BUG: Fix a non-constant expression used as structure initializer.

charris opened the pull request:
PyObject_HashNotImplemented was being used to initialize the tp_hash
slot in the PyArray_Type structure. In mingw32 (gcc version 3.4) that
results in a compile error. The function initialization seems to work
for more recent compilers, as the determination of what is considered
'non-constant' is left to the vendor and has changed.

charris committed:
BUG: Fix a non-constant expression used as structure initializer.

PyObject_HashNotImplemented was being used to initialize the tp_hash
slot in the PyArray_Type structure. In mingw32 (gcc version 3.4) that
results in a compile error. Fix by initializing to zero, then
setting to PyObject_HashNotImplemented when the module is loaded.

The function initialization seems to work with more recent compilers, as
the determination of what is considered 'non-constant' is left to the
vendor and has changed.

njsmith commented:
This fix looks very wrong to me, unless I misunderstand. Doesn't it have the effect that instead of `hash(array)` raising an error (which is what we want), then it will inherit the default `object.__hash__` so `hash(array)` will seem to succeed?

(Also if that's right and the tests pass then that is also a bug, because we should have a test for what I just said :-).)

charris commented:
It just reverts to the 1.9 version, so even if not optimal we have managed to live with it.

charris commented:
@njsmith Updated to initialize on load.

njsmith commented:
Two quibbles:
* it'd be good to have a comment next to the 0 saying "this is not real, it gets overwritten", because I know I at least already spend more time than I'd like trying to track down slot implementations, and I assume I'm not the only one!
* we should have a test checking that `hash(array)` fails... technically not this PR's responsibility, but this is a natural moment to add that.

Up to you on the latter really.

charris committed:
TST: Test that arrays are not hashable.

This adds a direct test to compliment the existing test that checks
if array is an instance of collections.Hashable.
